functions.inc.php

```diff
@@ -1918,4 +1918,21 @@ function getRemoteAddr() {
     return $REMOTE_ADDR;
 }
 
+
+/**
+ * Returns a hash for a username valid for one day
+ *
+ * @param String $username user name
+ * @return String password recovery code
+ */
+function getPasswordRecoveryCode($username)
+{
+    $username = trim(strtolower($username));
+    $date = date('Y-m-d');
+
+    $code = substr(strtoupper(md5('SECRET SALTING PHRASE' . $username . $date)), 0, 6);
```

Comment:
Using a fixed string + the mail address + the date (even without time) makes the recovery code easily computable and would directly translate into a CVE once we accept this code. (In other words: if you already run this code on a server already, disable it _now_.)

You'll need to add some randomness here, which also means you'll need to store the recovery code and the "valid until" date/time in the database.

Ideally we could use random_bytes(), but it was only added in PHP 7. So you'll need an `if function_exists('random_bytes')' switch with a fallback to openssl_random_pseudo_bytes(). I'd include at least 8 bytes of randomness. If none of these functions exists, throw an error. (Checking for them in setup.php might be a good idea, but it should be a warning only since it only affects an optional feature.)

Also, please include time() in the hash (not really a security improvement, but it can't hurt), and use sha1() instead of md5().

Finally, 6 chars are too short for a password recovery code - with the limited amount of characters used in a md5 or sha1 sum, it's too easy to guess. Please _at least_ make it 10 chars long - longer is better, and if someone forgets his password, he should at least suffer a bit when typing the code ;-)